pkg/apiserver/middlewares.go

```diff
@@ -0,0 +1,105 @@
+package apiserver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"go.creack.net/telepilot/pkg/jobmanager"
+)
+
+// Common method to extract the user's CN from context.
+func getUserFromContext(ctx context.Context) (string, error) {
+	var user string
+
+	if p, ok := peer.FromContext(ctx); ok {
+		if mtls, ok := p.AuthInfo.(credentials.TLSInfo); ok {
+			if len(mtls.State.PeerCertificates) > 1 {
+				return "", fmt.Errorf("too many peers in cert: %w", ErrInvalidClientCerts)
+			}
+			for _, item := range mtls.State.PeerCertificates {
+				user = item.Subject.CommonName
+				break
+			}
+		}
+	}
+	if user == "" {
+		return "", fmt.Errorf("CN not found: %w", ErrInvalidClientCerts)
+	}
+	return user, nil
+}
+
+// middleware to enforce authorization policies.
+func (s *Server) authMiddleware(user, fullMethod string, req any) error {
+	var j *jobmanager.Job
+	if getter, ok := req.(interface{ GetJobId() string }); ok {
+		jobID, err := uuid.Parse(getter.GetJobId())
+		if err != nil {
+			return fmt.Errorf("invalid uuid %q: %w", getter.GetJobId(), err)
+		}
+		job, err := s.jobmanager.LookupJob(jobID)
+		if err != nil {
+			return fmt.Errorf("lookup job %q: %w", getter.GetJobId(), err)
```

Comment:
Won't this reveal a difference between a job not found and a job that exists but the user doesn't have permission to operate on?

Reply:
Indeed. Made it a permission denied as well and added a test case to cover the scenario.